src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs

```diff
@@ -0,0 +1,433 @@
+                }
+
+                existing = _dictionary[propertyName];
+                if (ReferenceEquals(existing, value))
+                {
+                    // Ignore if the same value.
+                    return null;
+                }
+            }
+
+            int i = FindValueIndex(propertyName);
+            if (i >= 0)
+            {
+                if (_dictionary != null)
+                {
+                    _dictionary[propertyName] = value;
+                }
+                else
+                {
+                    KeyValuePair<string, T?> current = _list[i];
+                    if (ReferenceEquals(current.Value, value))
+                    {
+                        // Ignore if the same value.
+                        return null;
+                    }
+
+                    existing = current.Value;
+                }
+
+                assignParent?.Invoke();
+                _list[i] = new KeyValuePair<string, T?>(propertyName, value);
+            }
+            else
+            {
+                assignParent?.Invoke();
+                _dictionary?.Add(propertyName, value);
+                _list.Add(new KeyValuePair<string, T?>(propertyName, value));
+                Debug.Assert(existing == null);
+            }
+
+            return existing;
+        }
+
+        private void AddValue(string propertyName, T? value)
+        {
+            if (IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException_NodeCollectionIsReadOnly();
+            }
+
+            CreateDictionaryIfThreshold();
+
+            if (_dictionary == null)
+            {
+                // Verify there are no duplicates before adding.
+                if (ContainsProperty(propertyName))
+                {
+                    ThrowHelper.ThrowArgumentException_DuplicateKey(propertyName);
+                }
+            }
+            else
+            {
+                _dictionary.Add(propertyName, value);
+            }
+
+            _list.Add(new KeyValuePair<string, T?>(propertyName, value));
+        }
+
+        private bool TryAddValue(string propertyName, T? value)
+        {
+            if (IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException_NodeCollectionIsReadOnly();
+            }
+
+            CreateDictionaryIfThreshold();
+
+            if (_dictionary == null)
+            {
+                // Verify there are no duplicates before adding.
+                if (ContainsProperty(propertyName))
+                {
+                    return false;
+                }
+            }
+            else
+            {
+               if (!JsonHelpers.TryAdd(_dictionary, propertyName, value))
+                {
+                    return false;
+                }
+            }
+
+            _list.Add(new KeyValuePair<string, T?>(propertyName, value));
+            return true;
+        }
+
+        public void CreateDictionaryIfThreshold()
```

Comment:
nit: feels incomplete

```suggestion
        public void CreateDictionaryIfThresholdMet()
```